src/Resources/contao/controllers/BackendSwitch.php

```diff
@@ -106,25 +108,30 @@ public function run()
 			// Switch user accounts
 			if ($blnCanSwitchUser)
 			{
-				// Remove old sessions
-				$this->Database->prepare("DELETE FROM tl_session WHERE tstamp<? OR hash=?")
-							   ->execute(($time - \Config::get('sessionTimeout')), $strHash);
-
 			   // Log in the front end user
 				if (\Input::post('user'))
 				{
-					$objUser = \MemberModel::findByUsername(\Input::post('user'));
+					// Authenticate the new FrontendUser at the Symfony firewall
+					\System::getContainer()->get('contao.security.frontend_preview_authenticator')->authenticateFrontendUser(\Input::post('user'));
 
-					// Check the allowed member groups
-					if ($objUser !== null && ($this->User->isAdmin || \count(array_intersect(\StringUtil::deserialize($objUser->groups, true), $this->User->amg)) > 0))
+					if ($session->has(\FrontendUser::SECURITY_SESSION_KEY) && ($token = unserialize($session->get(\FrontendUser::SECURITY_SESSION_KEY))) instanceof TokenInterface)
 					{
-						// Insert the new session
-						$this->Database->prepare("INSERT INTO tl_session (pid, tstamp, name, sessionID, ip, hash) VALUES (?, ?, ?, ?, ?, ?)")
-									   ->execute($objUser->id, $time, 'FE_USER_AUTH', \System::getContainer()->get('session')->getId(), \Environment::get('ip'), $strHash);
-
-						// Set the cookie
-						$this->setCookie('FE_USER_AUTH', $strHash, ($time + \Config::get('sessionTimeout')), null, null, \Environment::get('ssl'), true);
-						$objTemplate->user = \Input::post('user');
+						/** @var User $user */
+						$user = $token->getUser();
+						$objUser = \MemberModel::findByUsername($user->getUsername());
```

Comment:
This is also something that makes me wonder. Shouldn't `$token->getUser()` return the user object already? Why are we loading the member model?

Reply:
The `getUser()` on the token can have a fully loaded user instance, but you cannot be sure.
Normally it holds only the unserialized data from the session storage.